encoding/protobuf/error.go

```diff
@@ -0,0 +1,174 @@
+		b.WriteString(" message:")
+		b.WriteString(err.message)
+	}
+	return b.String()
+}
+
+// NewError returns a new YARPC protobuf error. To access the error's fields, use the
+// provided `GetError*` methods. These methods are nil-safe and
+// provide default values for non-YARPC-errors.
+//
+// If the Code is CodeOK, this will return nil.
+func NewError(code yarpcerrors.Code, message string, options ...ErrorOption) error {
+	if code == yarpcerrors.CodeOK {
+		return nil
+	}
+	pbErr := &pberror{
+		code:    code,
+		message: message,
+	}
+	for _, opt := range options {
+		opt.apply(pbErr)
+	}
+	return pbErr
+}
+
+// GetErrorCode returns the error code of the error.
+func GetErrorCode(err error) yarpcerrors.Code {
+	if err == nil {
+		return yarpcerrors.CodeOK
+	}
+
+	if pberr, ok := err.(*pberror); ok {
+		return pberr.code
+	}
+	return yarpcerrors.CodeUnknown
+}
+
+// GetErrorMessage returns the error message of the error.
+func GetErrorMessage(err error) string {
+	if err == nil {
+		return ""
+	}
+
+	if pberr, ok := err.(*pberror); ok {
+		return pberr.message
+	}
+	return err.Error()
+}
+
+// GetErrorDetails returns the error details of the error.
+func GetErrorDetails(err error) []proto.Message {
+	if err == nil {
+		return nil
+	}
+
+	if pberr, ok := err.(*pberror); ok {
+		return pberr.details
+	}
+	return nil
+}
+
+// ErrorOption is an option for the NewError constructor.
+type ErrorOption struct{ apply func(*pberror) }
+
+// WithErrorDetails sets the details of the error.
+func WithErrorDetails(details ...proto.Message) ErrorOption {
+	return ErrorOption{func(err *pberror) {
+		err.details = append(err.details, details...)
+	}}
+}
+
+func convertToYARPCError(encoding transport.Encoding, err error) error {
+	if err == nil {
+		return nil
+	}
+	if pberr, ok := err.(*pberror); ok {
+		st, convertErr := status.New(_codeToGRPCCode[pberr.code], pberr.message).WithDetails(pberr.details...)
+		if convertErr != nil {
+			return convertErr
+		}
+		detailsBytes, cleanup, marshalErr := marshal(encoding, st.Proto())
+		defer cleanup()
+		if marshalErr != nil {
+			return marshalErr
+		}
+		yarpcDet := make([]byte, len(detailsBytes))
+		copy(yarpcDet, detailsBytes)
+		return yarpcerrors.Newf(pberr.code, pberr.message).WithDetails(yarpcDet)
+	}
+	return err
+}
+
+func convertFromYARPCError(encoding transport.Encoding, err error) error {
+	if err == nil || !yarpcerrors.IsStatus(err) {
+		return err
+	}
+	yarpcErr := yarpcerrors.FromError(err)
+	if yarpcErr.Details() == nil {
+		return err
+	}
+	st := &rpc.Status{}
+	unmarshalErr := unmarshalBytes(encoding, yarpcErr.Details(), st)
+	if unmarshalErr != nil {
+		return unmarshalErr
+	}
+
+	details := status.FromProto(st).Details()
+	var protobufDetails []proto.Message
+	var detailsErrs []error
+	for _, detail := range details {
+		if protoMessage, ok := detail.(proto.Message); ok {
+			protobufDetails = append(protobufDetails, protoMessage)
+		}
+		if detailsErr, ok := detail.(error); ok {
```

Comment:
Talked offline with @abhinav / @mh-park. We cannot return the failed marshal errors here; a server that returns a new error detail could break all of its clients, since the clients may not have been compiled with the Protobuf type.

That being said, there's not much value in separating `error`s and `proto.Messages` since users must attempt to type cast against their known types, (note: even if all Protobuf messages fail to decode, users will still get an error with a code and message). Users would likely only be concerned with the types they know about, and handle a default case for "future" details.

So unfortunately, this means we're probable best off aligning ourselves with Protobuf and have our `yarpcprotobuf.Deatils(err)` API return what the Protobuf API returns: a slice of interfaces :/ co-mingled with Protobuf message interfaces and error interfaces.